Avoid partial munmap memory leak

brandonpayton opened the pull request:
## What is this PR doing?

This PR works around a PHP memory leak that can quickly lead to an OOM condition.

## What problem is it solving?

This PR fixes #1128 -- an OOM condition due to memory leak.

## How is the problem addressed?

This PR adds custom PHP memory storage handlers to stop PHP from attempting to partial unmap memory via the `munmap()` function. Emscripten allocators do not support using `munmap()` to partially unmap memory. They fail for partial unmaps, and when attempts fail, memory is leaked because PHP thinks the memory is free while Emscripten libs consider it allocated.

The custom memory storage handlers take over responsibility for allocating and freeing aligned memory chunks. The handlers use `posix_memalign()` and `free()` instead of `mmap()` and `munmap()`, and this appears to resolve the problem.

The handlers are added as part of a PHP extension, which is not perfect because allocation can happen before the extension is initialized. But the amount of allocation before the workaround is applied should be quite small.

## Testing Instructions

Before applying this patch:
Run [this test script](https://gist.github.com/brandonpayton/934d96d76b96557e94ad6983aaffa09f) and verify that it OOMs.

After applying this patch:
Rerun the test script and verify that it completes after 1000 interations.

## Remaining work

- [x] Compare performance with php-wasm builds prior to this patch and reconsider if 20% slower.
- [x] Find out whether it is sufficient to add the workaround during the PHP module init phase rather than during each request init
- [x] Address minor TODO comments in wasm_memory_storage.c
- [x] See if we can run the test script as part of tests under `packages/php-wasm/node/src/test`

brandonpayton commented:
I rebased locally and am rebuilding php-wasm for node and web.

brandonpayton commented:
Updated the PR description to note that we should add the test script to the php-wasm node tests.

brandonpayton commented:
Please hold off testing this. Something is off here with the latest builds. The wasm_memory_storage extension is not present.

I verified this earlier before rebuilding for all PHP versions, but perhaps I did not test what I thought I tested. This should be a straightforward fix, but it's unfortunate others cannot yet try the fix for themselves. Planning to look into this first-thing in the morning.

brandonpayton commented:
Found the issue:
Somehow the Dockerfile updates in this PR were missing the COPY instruction that I added earlier today for copying in the wasm_memory_storage extension into php-src.

I'm not yet sure what might have happened for that to be omitted. Instead of building all PHP versions again tonight, I'll go ahead and build just light PHP 7.0 and PHP 8.3 for the web.

One odd thing I noticed when testing after rebasing:
The OOM test script output is no longer directed to the browser when there is an OOM. Instead there is just an error in the dev tools console with the OOM message. This seems undesirable as a user may prefer some output indicating breakage than hidden output in the case of breakage. Will make a note to confirm that behavior tomorrow.

brandonpayton commented:
If others are interested, the build is fixed, and this is now ready for exploratory testing.

I've included just the light php-wasm builds for the web for PHP 7.0 and PHP 8.3 for now.

dmsnell commented on packages/php-wasm/compile/php-wasm-memory-storage/php_wasm_memory_storage.h:
```diff
@@ -0,0 +1,11 @@
+extern zend_module_entry wasm_memory_storage_module_entry;
+# define phpext_wasm_memory_storage_ptr &wasm_memory_storage_module_entry
+
+# define PHP_WASM_MEMORY_STORAGE_VERSION "0.0.1"
```
curious since I'm unfamiliar: why the space between the `#` and `define`?

adamziel replied:
TIL you can have a space there

brandonpayton replied:
This is just code that was generated by:

https://github.com/php/php-src/blob/master/ext/ext_skel.php

I found the extra space curious too.

dmsnell commented on packages/php-wasm/compile/php-wasm-memory-storage/wasm_memory_storage.c:
```diff
@@ -0,0 +1,87 @@
+void* wasm_memory_storage_chunk_alloc(zend_mm_storage* storage, size_t size, size_t alignment)
+{
+	void* ptr = NULL;
+	if (posix_memalign(&ptr, alignment, size) == 0)
```
> posix_memalign() should be used judiciously as the algorithm that
> realizes the alignment constraint can incur significant memory overhead.

from the `man` page. I'm curious here if they are referring to the waste of unaligned memory that _could_ be allocated or to something else, something worse

adamziel replied:
+1

brandonpayton replied:
@dmsnell which version of the man page?

In the man pages I looked at on the web, I'm not seeing such a word of caution. I hope that is a hint that it is probably OK.

It would be interesting to read how Emscripten allocators achieve the alignment. I can skim that later today for any hint of additional storage requirements. But I don't see why there would need to be. It is just allocated memory that can be tracked the same way as everything else, I would think. ...Unless there is memory used to make future aligned allocations faster or something.

dmsnell replied:
@brandonpayton both the `man 3 posix_memalign` page on my macOS install plus the one downloaded into my docs browser share that message in the **Notes** section.

kind of makes me wonder if the posix version isn't doing something similar to what PHP does. maybe it's not faster? maybe it's just automatic.

brandonpayton replied:
👍   I was mobile before and was just searching. Hadn't checked the macOS one yet :)  Thank you for bringing this up.

Regarding speed, I did some semi-naive perf testing, and with 1000 sequential requests, it took 8 seconds with this change and 9 seconds without it. I can share details later in a separate comment.

dmsnell replied:
sounds close enough that it doesn't matter

brandonpayton replied:
> kind of makes me wonder if the posix version isn't doing something similar to what PHP does. maybe it's not faster? maybe it's just automatic.

That is indeed how Emscripten's default allocator works:
https://github.com/emscripten-core/emscripten/blob/main/system/lib/dlmalloc.c#L5001-L5047

It asks for enough memory to guarantee an aligned address followed by at least enough memory. Then it gives back the unused head and tail.

I still think it is better to be explicit with a memalign() call than to manually do this with mmap() and munmap() because
1. A memalign() call more precisely expresses intent.
2. A memalign() call gives libc and/or the OS the opportunity to satisfy the request more efficiently, while using mmap()/munmap() ties you to a specific approach.

adamziel commented:
I've played with performance measurements.

Methodology:

* Close all other Playground tabs and open a single new tab
* Open devtools before opening the page
* Go directly to a local Playground URL that loads the respective PHP version and WP 6.5
* Paste the code snippet below into devtools
* Don't touch the computer until the test is finished. In particular, don't interact with the page and don't switch to another browser tab.

These may seem like insignificant details, but Chrome does very different things under the hood with and without devtools (different WASM runtimes) and also if the existing tab was used before (Playwright developers struggled with intermittent failures when reusing tabs).

```ts
async function measureTime() {
    console.time("ExecutionTime"); // Start the timer
    for(let i = 0; i < 100; i++) {
        await playground.request({ url: '/' });
        if ((i+1) % 10 === 0) { // Log progress every 10 requests
            console.log(`Progress: Completed ${i+1} out of 100 requests`);
        }
    }
    console.timeEnd("ExecutionTime"); // End the timer and print the total time taken
}
measureTime();
```

Here are the results of 4 test runs on each version:

* PHP 8.3 from this PR: `29 765 ms`,  `29 112ms`, `29 126 ms`, `29 180ms`
* PHP 8.3 from trunk: `29 701 ms`, `29 334ms`, `29 383ms`, `29 317ms`

My testing methodology requires some manual work and I won't spend the time to collect 30 or 500 samples to plot a distribution and calculate p-values. We could automate that with a Node.js build, but at this point I'm quite happy to assume we're not looking at a shattering performance decline. I've [captured this comment in the project wiki](https://github.com/WordPress/wordpress-playground/wiki/Performance-testing-resources) for any future performance discussions.

For now – **let's assume the performance is alright here and proceed with code review as usual** 👍 

Even if there's a much slower code branch somewhere, I'd rather fix five more crashes than obsess about all the performance implications – we'll focus on the speed more if and when that matters.

Thank you so much for all your work on this @brandonpayton!

adamziel commented on packages/php-wasm/compile/php-wasm-memory-storage/wasm_memory_storage.c:
```diff
@@ -0,0 +1,87 @@
+/* wasm_memory_storage extension for PHP */
```
Let's also link to this PR and the related issue to give future readers more context

brandonpayton commented:
> I've played with performance measurements.
> 
> Methodology:
> [snip]
> These may seem like insignificant details, but Chrome does very different things under the hood with and without devtools (different WASM runtimes) and also if the existing tab was used before (Playwright developers struggled with intermittent failures when reusing tabs).

@adamziel Thanks for testing and for sharing the clear methodology.

I ran some tests but with less care about creating a new tab and opening dev tools first. I didn't remember about console.time() and didn't know about playground.request(). Here's the script I used (with the current scope URL substituted):
```js
async function* test100() {
    for (let i = 0; i < 100; ++i) {
        yield await fetch( '/scope:0.7238152673090998/' ).then( r => r.text() );
    }
}

let start = performance.now();
console.log( 'Start:', start );
for await ( let t of test100() ) {}
let end = performance.now();
console.log( 'End:', end );
console.log( 'Total:', end - start );
```

There isn't a reason that the response text was needed, but originally, I was thinking about logging part of it.

Without this PR, the total time was about 8.97 seconds.
With this PR, the total time was about 7.87 seconds.

I prefer your more careful tests but wanted to share this sample.

brandonpayton commented:
@dmsnell, I added more documentation, including links to inline examples of custom storage functions, but so far, I have not been able to find related specs or external documentation to link to. Please feel free to ask for more if you think something is missing.

@adamziel, I was able to move the workaround assignment into module init. Zend memory manager does do some [cleanup during request shutdown](https://github.com/php/php-src/blob/dbaeb62ab1e34067057170ab50cf39d1bde584d8/main/main.c#L1918), but it only creates its heap during module startup.

It is possible other modules may have allocated memory before the workaround is added, and we should give those a chance to free the memory in the same way it was allocated. Fortunately, when PHP chooses to run module shutdown, it appears to do so in reverse order which gives us a chance to remove the workaround before those initial allocations are freed. So I added a module shutdown function to cleanup after ourselves.

With the exception of adding a test for the OOM case, this PR is ready for another review.

adamziel commented:
This looks great @brandonpayton! I'm happy to merge as soon as all the PHP versions are rebuilt.

Should we now be strict about using `emalloc` and `efree` in php_wasm.c? Or is `malloc` and `free` still fine?

brandonpayton commented:
I added a test for this fix that runs for all supported PHP versions:
https://github.com/WordPress/wordpress-playground/pull/1189/commits/2e370376c3172f6508033a39903156dac26dd23b

All these tests fail with an OOM error before rebuilding php-wasm with the fix. After the fix, the tests pass.

This is ready for another review.

brandonpayton commented:
> This looks great @brandonpayton! I'm happy to merge as soon as all the PHP versions are rebuilt.

@adamziel, sounds good! I will rebuild the PHP versions but leave for merging tomorrow, in case there is some auto-deploy to playground.wordpress.net and there are unexpected problems during the night.

> Should we now be strict about using `emalloc` and `efree` in php_wasm.c? Or is `malloc` and `free` still fine?

At a high level, I guess `emalloc` and `efree` should be used if you want PHP's memory manager involved. But based on my current understanding, it should be safe to use either pair, as long as we don't mix them and attempt to `efree()` something allocated with `malloc()`.

brandonpayton commented:
Note: I am stepping away for the night but will leave a build command running that should commit and push when complete:
```bash
npm run "recompile:php:node" && npm run "recompile:php:web" && git add -u && git commit -m "Add php-wasm builds" && git push --force-with-lease
```

dmsnell reviewed (approved):
Looks great, and thanks for the comments. Too bad there's not more clarity in PHP around creating these custom internals, but the links to a versioned source code file are about as perfect as I suppose we can get.

Looking forward to how this goes. It will make many people happy.

brandonpayton commented:
With trying multiple things within this PR, especially with some of the force-pushing that rewrote previous history and hid associated test runs, it turned out to be very hard for anyone to really follow all the things I've tried so far. And it was even hard for me to follow because links to previous CI runs were hidden after force push.

I think we should probably close this and use dedicated PRs for each of the approaches. We can leave multiple investigatory PRs open at the same time and just keep documenting the findings under the original issue #1128.

For example, the following would be separate PRs:
- Instrumenting PHP and JS code to learn more about the crash
- Checking whether upgrading Emscripten seems to work around the issue or adds better checks that help identify the problem
- Enabling ASSERTIONS and SAFE_HEAP to see whether they provide more info
- Using our own pre-build version of node that will print the actual out of bounds address

I am working on creating more focused PRs to help nail this problem down with less confusion.
